New Tool PR : ImageGenerator Dall-E Backend

void-mckenzie opened the pull request:
# New Tool PR : ImageGenerator
**Tool Name**  : ImageGenerator

**Tool Description** : Tool for generating an image from a given prompt using DALL·E API backend.

```
@activity(
        config={
            "description": "Generates single image using DALL·E API based on a provided prompt and returns the image data in bytes if the mode is 'download' or returns the URL of the image if the mode is 'url' ",
            "schema": Schema(
                {
                    "prompt": str,
                    Optional("size", default="1024x1024"): str,
                    Optional("mode", default="url"): str,
                }
            ),
        }
    ) 
```

**List of Proposed Activities**
1. Takes in a prompt (required parameter), an optional 'size' parameter and an optional 'mode' parameter from the user/agent and calls the Dall-E API to generate an image.

2. The Dall-E API takes in the prompt and size, returning the URL of the generated image to the tool.

3. The tool, depending on the 'mode', either:
    - [mode = 'url'] returns the URL to the agent (expected behavior akin to langchain's image generator)
    - [mode = 'download'] uses requests library to get the image data and returns it as a BlobArtifact to the agent. It is stored in the ToolMemory. Griptape's FileManager tool can use it to save/download the image to the disk.

## Examples with Griptape Agent

**Code template for testing the tool with an agent:**

```
from griptape.structures import Agent
from griptape.tools import FileManager, ImageGenerator

fm_tool = FileManager()
dalle_tool = ImageGenerator()


agent = Agent(
    tools=[dalle_tool, fm_tool]
)

output = agent.run(
    "Enter Prompt here"
)

```

**1. URL of 1 image**

<ins>Prompt</ins>:      "gothic mansion with ivy growing on the walls, size 512x512."    

<ins>Agent Result</ins>:

```
[10/26/23 11:08:16] INFO     ToolkitTask 5f61076b6a0b4190b641e49e91b12cde      
                             Input: gothic mansion with ivy growing on the     
                             walls, size 512x512.                              
[10/26/23 11:08:34] INFO     Subtask 9fc64ffb34324fc7a38ee16f21c507e3          
                             Thought: The user wants an image of a gothic      
                             mansion with ivy growing on the walls. I can use  
                             the ImageGenerator tool with the dalle_generate   
                             activity to create this image. The size of the    
                             image should be 512x512.                          
                                                                               
                             Action: {"type": "tool", "name": "ImageGenerator",
                             "activity": "dalle_generate", "input": {"values": 
                             {"prompt": "gothic mansion with ivy growing on the
                             walls", "size": "512x512", "mode": "url"}}}       
[10/26/23 11:08:42] INFO     Subtask 9fc64ffb34324fc7a38ee16f21c507e3          
                             Observation: Output of                            
                             "ImageGenerator.dalle_generate" was stored in     
                             memory with memory_name "ToolMemory" and          
                             artifact_namespace                                
                             "1e388a7785ef4832a4921e4767c77eba"                
[10/26/23 11:08:53] INFO     Subtask ffba4d4c52324f198fa6615c93d175f6          
                             Thought: The image has been generated and stored  
                             in memory. I need to retrieve it from memory to   
                             provide the URL to the user. I can use the        
                             ToolMemory with the query activity to get this    
                             information.                                      
                             Action: {"type": "memory", "name": "ToolMemory",  
                             "activity": "query", "input": {"values":          
                             {"memory_name": "ToolMemory",                     
                             "artifact_namespace":                             
                             "1e388a7785ef4832a4921e4767c77eba", "query": "What
                             is the URL of the generated image?"}}}            
[10/26/23 11:09:28] INFO     Subtask ffba4d4c52324f198fa6615c93d175f6          
                             Observation: The URL of the generated image is    
                             https://oaidalleapiprodscus.blob.core.windows.net/
                             private/org-KTLk70H0J9NHECwZRkcWwXiJ/user-4hXSDe9K
                             XuBUSAoWgmwrq20i/img-o7vtIwf1jpNdj5STQn2XDJh6.png?
                             st=2023-10-26T17%3A08%3A41Z&se=2023-10-26T19%3A08%
                             3A41Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=ima
                             ge/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&
                             sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=202
                             3-10-25T22%3A26%3A59Z&ske=2023-10-26T22%3A26%3A59Z
                             &sks=b&skv=2021-08-06&sig=eASr//UFJ2MB84tqYGx9TnvB
                             hoAwRmGFczF5W3Iyidc%3D                            
[10/26/23 11:10:05] INFO     ToolkitTask 5f61076b6a0b4190b641e49e91b12cde      
                             Output: Here is the image of a gothic mansion with
                             ivy growing on the walls: [Image                  
                             URL](https://oaidalleapiprodscus.blob.core.windows
                             .net/private/org-KTLk70H0J9NHECwZRkcWwXiJ/user-4hX
                             SDe9KXuBUSAoWgmwrq20i/img-o7vtIwf1jpNdj5STQn2XDJh6
                             .png?st=2023-10-26T17%3A08%3A41Z&se=2023-10-26T19%
                             3A08%3A41Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsc
                             t=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786
                             b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&sk
                             t=2023-10-25T22%3A26%3A59Z&ske=2023-10-26T22%3A26%
                             3A59Z&sks=b&skv=2021-08-06&sig=eASr//UFJ2MB84tqYGx
                             9TnvBhoAwRmGFczF5W3Iyidc%3D)                      

```

**2. Download 1 image**

<ins>Prompt</ins>:      "I want to download an image of a gothic mansion with ivy growing on the walls. Save it in the current directory"        
  

<ins>Agent Result</ins>:

```
[10/26/23 10:38:23] INFO     ToolkitTask b1778c191d754b569704a31d9c12ef1c      
                             Input: I want to download an image of a gothic    
                             mansion with ivy growing on the walls. Save it in 
                             the current directory                             
[10/26/23 10:38:44] INFO     Subtask 1803993dc0244bfd860ffdaa08b1ddd1          
                             Thought: To fulfill the user's request, I need to 
                             generate an image using the ImageGenerator tool   
                             with the prompt "gothic mansion with ivy growing  
                             on the walls". Then, I need to save the image to  
                             the current directory using the FileManager tool. 
                                                                               
                             Action: {"type": "tool", "name": "ImageGenerator",
                             "activity": "dalle_generate", "input": {"values": 
                             {"prompt": "gothic mansion with ivy growing on the
                             walls", "mode": "download"}}}                     
[10/26/23 10:38:54] INFO     Subtask 1803993dc0244bfd860ffdaa08b1ddd1          
                             Observation: Output of                            
                             "ImageGenerator.dalle_generate" was stored in     
                             memory with memory_name "ToolMemory" and          
                             artifact_namespace                                
                             "5047ccdda18b4c53bbcde6b29cb8dab4"                
[10/26/23 10:39:18] INFO     Subtask 1dc8321c60b94570a02e7d5e1f3b9322          
                             Thought: Now that the image has been generated and
                             stored in memory, I need to save it to the current
                             directory. I will use the FileManager tool's      
                             save_memory_artifacts_to_disk activity for this. I
                             will save it in the current directory, so I will  
                             use '.' as the directory name. I will give the    
                             file a name that reflects the content of the      
                             image.                                            
                             Action: {"type": "tool", "name": "FileManager",   
                             "activity": "save_memory_artifacts_to_disk",      
                             "input": {"values": {"dir_name": ".", "file_name":
                             "gothic_mansion_with_ivy.jpg", "memory_name":     
                             "ToolMemory", "artifact_namespace":               
                             "5047ccdda18b4c53bbcde6b29cb8dab4"}}}             
                    INFO     Subtask 1dc8321c60b94570a02e7d5e1f3b9322          
                             Observation: saved successfully                   
[10/26/23 10:39:24] INFO     ToolkitTask b1778c191d754b569704a31d9c12ef1c      
                             Output: The image of a gothic mansion with ivy    
                             growing on the walls has been successfully        
                             downloaded and saved in the current directory with
                             the filename "gothic_mansion_with_ivy.jpg".       
```

**3. Download multiple images**

<ins>Prompt</ins>:      "I want to download 2 images of gothic mansions with ivy growing on the walls."        
  

<ins>Agent Result</ins>:

```
[10/26/23 10:46:32] INFO     ToolkitTask 7d4bd6419d0a464e816271e8ccf55d44      
                             Input: I want to download 2 images of gothic      
                             mansions with ivy growing on the walls.           
[10/26/23 10:46:52] INFO     Subtask 55028143c0bd4b1b815b97196e29093c          
                             Thought: The user wants to download two images of 
                             gothic mansions with ivy growing on the walls. I  
                             can use the ImageGenerator tool with the          
                             dalle_generate activity to generate these images. 
                             I'll start by generating the first image.         
                                                                               
                             Action: {"type": "tool", "name": "ImageGenerator",
                             "activity": "dalle_generate", "input": {"values": 
                             {"prompt": "Gothic mansion with ivy growing on the
                             walls", "mode": "download"}}}                     
[10/26/23 10:47:02] INFO     Subtask 55028143c0bd4b1b815b97196e29093c          
                             Observation: Output of                            
                             "ImageGenerator.dalle_generate" was stored in     
                             memory with memory_name "ToolMemory" and          
                             artifact_namespace                                
                             "5e525b753a3a4661aaa14a3cfbb60503"                
[10/26/23 10:47:24] INFO     Subtask c14c009e1bca4ce9a7a2182f019115a0          
                             Thought: The first image has been generated and   
                             stored in memory. Now, I need to save this image  
                             to disk using the FileManager tool with the       
                             save_memory_artifacts_to_disk activity. I'll      
                             specify a directory and filename for the image.   
                             Action: {"type": "tool", "name": "FileManager",   
                             "activity": "save_memory_artifacts_to_disk",      
                             "input": {"values": {"dir_name": "images",        
                             "file_name": "gothic_mansion_1.jpg",              
                             "memory_name": "ToolMemory", "artifact_namespace":
                             "5e525b753a3a4661aaa14a3cfbb60503"}}}             
                    INFO     Subtask c14c009e1bca4ce9a7a2182f019115a0          
                             Observation: saved successfully                   
[10/26/23 10:47:36] INFO     Subtask 47f94399f9a24b5f9f11214846982eed          
                             Thought: The first image has been successfully    
                             saved to disk. Now, I need to generate the second 
                             image using the ImageGenerator tool with the      
                             dalle_generate activity.                          
                             Action: {"type": "tool", "name": "ImageGenerator",
                             "activity": "dalle_generate", "input": {"values": 
                             {"prompt": "Gothic mansion with ivy growing on the
                             walls", "mode": "download"}}}                     
[10/26/23 10:47:45] INFO     Subtask 47f94399f9a24b5f9f11214846982eed          
                             Observation: Output of                            
                             "ImageGenerator.dalle_generate" was stored in     
                             memory with memory_name "ToolMemory" and          
                             artifact_namespace                                
                             "a6e943b6b83845a79773aaa53f64492a"                
[10/26/23 10:48:05] INFO     Subtask fe38ed34e1664291b10d1b3797a79174          
                             Thought: The second image has been generated and  
                             stored in memory. Now, I need to save this image  
                             to disk using the FileManager tool with the       
                             save_memory_artifacts_to_disk activity. I'll      
                             specify a directory and filename for the image.   
                             Action: {"type": "tool", "name": "FileManager",   
                             "activity": "save_memory_artifacts_to_disk",      
                             "input": {"values": {"dir_name": "images",        
                             "file_name": "gothic_mansion_2.jpg",              
                             "memory_name": "ToolMemory", "artifact_namespace":
                             "a6e943b6b83845a79773aaa53f64492a"}}}             
                    INFO     Subtask fe38ed34e1664291b10d1b3797a79174          
                             Observation: saved successfully                   
[10/26/23 10:48:10] INFO     ToolkitTask 7d4bd6419d0a464e816271e8ccf55d44      
                             Output: The two images of gothic mansions with ivy
                             growing on the walls have been successfully       
                             downloaded and saved as "gothic_mansion_1.jpg" and
                             "gothic_mansion_2.jpg" in the "images" directory. 
```

**4. Multiple image URLs**

<ins>Prompt</ins>:      "I want 2 images of gothic mansions with ivy growing on the walls."        
  

<ins>Agent Result</ins>:

```
[10/26/23 11:03:42] INFO     ToolkitTask abcbfe00aed64fa082501bf94b56f397      
                             Input: I want 2 images of gothic mansions with ivy
                             growing on the walls.                             
[10/26/23 11:04:00] INFO     Subtask 4457070f7f68436da018c8e430d834f3          
                             Thought: The user wants two images of gothic      
                             mansions with ivy growing on the walls. I can use 
                             the ImageGenerator tool with the "dalle_generate" 
                             activity to generate these images. I will need to 
                             do this twice to get two images.                  
                                                                               
                             Action: {"type": "tool", "name": "ImageGenerator",
                             "activity": "dalle_generate", "input": {"values": 
                             {"prompt": "Gothic mansion with ivy growing on the
                             walls", "size": "1024x1024", "mode": "url"}}}     
[10/26/23 11:04:08] INFO     Subtask 4457070f7f68436da018c8e430d834f3          
                             Observation: Output of                            
                             "ImageGenerator.dalle_generate" was stored in     
                             memory with memory_name "ToolMemory" and          
                             artifact_namespace                                
                             "aa5a19d9ce454778b0889d02083bbb6f"                
[10/26/23 11:04:19] INFO     Subtask df808aab0690411f890cf7b578089b24          
                             Thought: The first image has been generated and   
                             stored in memory. Now I need to generate the      
                             second image using the same process.              
                             Action: {"type": "tool", "name": "ImageGenerator",
                             "activity": "dalle_generate", "input": {"values": 
                             {"prompt": "Gothic mansion with ivy growing on the
                             walls", "size": "1024x1024", "mode": "url"}}}     
[10/26/23 11:04:27] INFO     Subtask df808aab0690411f890cf7b578089b24          
                             Observation: Output of                            
                             "ImageGenerator.dalle_generate" was stored in     
                             memory with memory_name "ToolMemory" and          
                             artifact_namespace                                
                             "352b4b4eeb98421fbec76450770276cc"                
[10/26/23 11:04:39] INFO     Subtask 1d4c8916c0414db480dc7f38a9296967          
                             Thought: Both images have been generated and      
                             stored in memory. Now I need to retrieve these    
                             images from memory using the "query" activity of  
                             the "ToolMemory".                                 
                             Action: {"type": "memory", "name": "ToolMemory",  
                             "activity": "query", "input": {"values":          
                             {"memory_name": "ToolMemory",                     
                             "artifact_namespace":                             
                             "aa5a19d9ce454778b0889d02083bbb6f", "query": "What
                             is the URL of the generated image?"}}}            
[10/26/23 11:05:13] INFO     Subtask 1d4c8916c0414db480dc7f38a9296967          
                             Observation: The URL of the generated image is    
                             https://oaidalleapiprodscus.blob.core.windows.net/
                             private/org-KTLk70H0J9NHECwZRkcWwXiJ/user-4hXSDe9K
                             XuBUSAoWgmwrq20i/img-atXj83TOcRKqFaHnJXslAzR0.png?
                             st=2023-10-26T17%3A04%3A08Z&se=2023-10-26T19%3A04%
                             3A08Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=ima
                             ge/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&
                             sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=202
                             3-10-26T13%3A59%3A00Z&ske=2023-10-27T13%3A59%3A00Z
                             &sks=b&skv=2021-08-06&sig=rD6aodAzUXuK7/veDzIhjWc0
                             hrikoXvYP3Twelr0AHI%3D                            
[10/26/23 11:05:22] INFO     Subtask 4353cd03dc324b9799a40a1988f65b36          
                             Thought: I have retrieved the URL of the first    
                             image. Now I need to retrieve the URL of the      
                             second image from memory.                         
                             Action: {"type": "memory", "name": "ToolMemory",  
                             "activity": "query", "input": {"values":          
                             {"memory_name": "ToolMemory",                     
                             "artifact_namespace":                             
                             "352b4b4eeb98421fbec76450770276cc", "query": "What
                             is the URL of the generated image?"}}}            
[10/26/23 11:05:58] INFO     Subtask 4353cd03dc324b9799a40a1988f65b36          
                             Observation: The URL of the generated image is    
                             https://oaidalleapiprodscus.blob.core.windows.net/
                             private/org-KTLk70H0J9NHECwZRkcWwXiJ/user-4hXSDe9K
                             XuBUSAoWgmwrq20i/img-jdq2UERDc28hvRaFTaLnB1Dp.png?
                             st=2023-10-26T17%3A04%3A27Z&se=2023-10-26T19%3A04%
                             3A27Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=ima
                             ge/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&
                             sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=202
                             3-10-26T13%3A20%3A47Z&ske=2023-10-27T13%3A20%3A47Z
                             &sks=b&skv=2021-08-06&sig=wtmYhjp5xhTjJMBnXVDGgTAB
                             liO2RTw5lkq7ePBfZ6M%3D                            
[10/26/23 11:07:00] INFO     ToolkitTask abcbfe00aed64fa082501bf94b56f397      
                             Output: Here are the two images of gothic mansions
                             with ivy growing on the walls:                    
                                                                               
                             1. [Image                                         
                             1](https://oaidalleapiprodscus.blob.core.windows.n
                             et/private/org-KTLk70H0J9NHECwZRkcWwXiJ/user-4hXSD
                             e9KXuBUSAoWgmwrq20i/img-atXj83TOcRKqFaHnJXslAzR0.p
                             ng?st=2023-10-26T17%3A04%3A08Z&se=2023-10-26T19%3A
                             04%3A08Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=
                             image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b0
                             67&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=
                             2023-10-26T13%3A59%3A00Z&ske=2023-10-27T13%3A59%3A
                             00Z&sks=b&skv=2021-08-06&sig=rD6aodAzUXuK7/veDzIhj
                             Wc0hrikoXvYP3Twelr0AHI%3D)                        
                             2. [Image                                         
                             2](https://oaidalleapiprodscus.blob.core.windows.n
                             et/private/org-KTLk70H0J9NHECwZRkcWwXiJ/user-4hXSD
                             e9KXuBUSAoWgmwrq20i/img-jdq2UERDc28hvRaFTaLnB1Dp.p
                             ng?st=2023-10-26T17%3A04%3A27Z&se=2023-10-26T19%3A
                             04%3A27Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=
                             image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b0
                             67&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=
                             2023-10-26T13%3A20%3A47Z&ske=2023-10-27T13%3A20%3A
                             47Z&sks=b&skv=2021-08-06&sig=wtmYhjp5xhTjJMBnXVDGg
                             TABliO2RTw5lkq7ePBfZ6M%3D)                        
```

andrewfrench commented on griptape/tools/image_gen/tool.py:
```diff
@@ -0,0 +1,46 @@
+import requests
+
+
+class ImageGenerator(BaseTool):
```
We should match the pattern throughout other tools and define an `attrs` class using the @attrs.define decorator here.

andrewfrench commented on griptape/tools/__init__.py:
```diff
@@ -44,4 +45,5 @@
     "Computer",
     "ProxycurlClient",
     "OpenWeatherClient",
+    "ImageGenerator",
```
I can imagine future support for other image generation providers beyond DALL-E. Can you rename this tool to something like DalleImageGenerator?

void-mckenzie replied:
Hey Andrew, got it! I just realized I haven't used error artifacts here either. I'll make the necessary changes and include a unit test file on my next commit. Will keep you posted. 
- Mukkesh

andrewfrench commented on griptape/tools/image_gen/tool.py:
```diff
@@ -0,0 +1,46 @@
+
+
+class ImageGenerator(BaseTool):
+    api_key = os.environ.get("OPENAI_API_KEY")
```
We should define a field here with a sensible default:

```suggestion
    api_key: Optional[str] = field(
        default=Factory(lambda: os.environ.get("OPENAI_API_KEY")), kw_only=True
    )
```

void-mckenzie replied:
Done!

andrewfrench reviewed (changes requested):
Very cool! A couple quick comments from me.

void-mckenzie commented:
Made all the changes @andrewfrench  :)

- Added attr define, field and factory to bring it to the griptape standard.
- Renamed the tool to DalleImageGenerator. 
- Ran Poetry Black on all the files to fix formatting issues. 
- Added a unit test file to assert errors as of now. 

The only issue was that Optional exists in both typing and schema. So I imported one of them as TypingOptional to prevent clashes.  
Passed all the checks. Please take a look at it when you get time and let me know if I need to make any more changes. Cheers!

andrewfrench commented:
Hi @void-mckenzie, and thanks again for the PR! After some internal discussion, we've decided to make some larger scale updates to the framework to support image generation generally rather than providing tools for individual models. This approach is outlined in #332.

As always, you can create custom tools or drivers and import them to be used with Griptape for your specific use cases, but we'll close this in preparation for generalized multi-model image generation support.
